Add new sample

h5aaimtron opened the pull request:
SignalR Core Chat Sample with Auth (JWT)

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/aspnet/SignalR-samples?pullRequest=89) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/aspnet/SignalR-samples?pullRequest=89) before we can accept your contribution.<br/><br/>:x: h5aaimtron [sign now](https://cla.dotnetfoundation.org/aspnet/SignalR-samples?pullRequest=89)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/aspnet/SignalR-samples?pullRequest=89) it.</sub>

bradygaster commented on ChatWithAuthSample/AuthorizationSample/Controllers/ValuesController.cs:
```diff
@@ -0,0 +1,45 @@
+﻿using System;
```
Why leave this file in if unused anywhere?

h5aaimtron replied:
Remnant that I can definitely delete.

bradygaster commented on ChatWithAuthSample/AuthorizationSample/Dockerfile:
```diff
@@ -0,0 +1,23 @@
+#Depending on the operating system of the host machines(s) that will build or run the containers, the image specified in the FROM statement may need to be changed.
+#For more information, please see https://aka.ms/containercompat
+
+FROM microsoft/dotnet:2.1-aspnetcore-runtime-nanoserver-1803 AS base
```
We're close to releasing .NET Core 3.0. Would you be interested in upgrading the app to 3.0 preview 7? I think that'd provide more value.

h5aaimtron replied:
I can give it a go this weekend and let you know if it all still works.

h5aaimtron replied:
I updated it to .NET Core 3.0 Preview 6 and checked it in. I started on the Angular app which I'll use to test and verify. Will get back to you all once I can confirm it is good to go.

h5aaimtron replied:
Tested/working with the update to .NET Core 3.0 Preview 6. Angular 7 App:
https://github.com/h5aaimtron/sample_chatWithIdentityClient

davidfowl replied:
Why is this a windows container

h5aaimtron replied:
Defaulted docker to windows at the time. I switch back and forth. I can update it as needed.

bradygaster commented on ChatWithAuthSample/AuthorizationSample/Hubs/ChatHub.cs:
```diff
@@ -0,0 +1,100 @@
+                // Add user to roster.
+                lock(UserRoster)
+                {
+                    UserRoster.Add(Context.ConnectionId, Context.User.Identity.Name);
```
Great example of matching a user to a connection. extra mile!

bradygaster commented on ChatWithAuthSample/AuthorizationSample/Startup.cs:
```diff
@@ -0,0 +1,177 @@
+                .AddDefaultTokenProviders()
+                .AddEntityFrameworkStores<Persistence.UserContext>();
+
+            services.Configure<IdentityOptions>(options =>
```
really comprehensive way to mix identity in. this is a great example.

h5aaimtron replied:
Thank you for your kind words, it is appreciated. I'm in the process of developing a very complex SignalR Core based app at the moment and decided to distill down some of the stuff I learned along the way. I'm going to be likely adding an Angular 6+ client repo for this sample this weekend as there seems to be interest in the issue tracker. I'm more than willing to add that into the sample work if you would like.

bradygaster commented on ChatWithAuthSample/AuthorizationSample/Startup.cs:
```diff
@@ -0,0 +1,177 @@
+            // Enable Authentication First
+            app.UseAuthentication();
+            // Set CORS Policy, allowing all right now, should be configured for needs.
+            app.UseCors(x => x.AllowAnyOrigin().AllowAnyMethod().AllowAnyHeader().AllowCredentials());
```
Have you validated this will work? Later in the .NET Core releases CORS became a bit more restrictive (a good thing), so it'd be good for you to test this out to make sure all the connections are working in both a localhost environment and once published. I think you have to specify at least an origin now.

h5aaimtron replied:
I've tested this both locally and remotely on a VPS I have hosted on 1and1.com. It is definitely not the preferred configuration, however; it alleviates development issues initially. I would always recommend restricting this, but wanted to keep it as simple as possible.

h5aaimtron replied:
Validated in .NET Core 3.0 Preview 6. Had to add localhost to the CORS per your note. Angular App Repo: https://github.com/h5aaimtron/sample_chatWithIdentityClient

h5aaimtron commented:
I updated all the code and have the repos ready. Let me know if you want to proceed to create specific repos for them and what/if any naming you'd prefer.

bradygaster commented:
Thanks for all that work over the weekend, @h5aaimtron. i'll clone your branch and give it a go and provided everything works just fine, I'd create an individual repo named `sample_AngularChat` (unless @davidfowl) has another idea in the SignalR org and stash it there.

bradygaster commented:
Also - It'd be great to have a page that's rendered for the user to actually use the API and Hub. From what I can gather in your fork, there's no UI or anything for a user to use. A README.md on how to try the sample out, or a UI making it easy for a user to "clone, build, and see" it working would be optimal. Is that possible given your scenario?

h5aaimtron commented:
@bradygaster, I put the UI in a separate repo here: https://github.com/h5aaimtron/sample_chatWithIdentityClient

It's a relatively basic Angular 6 app. I wasn't sure if/how you'd like them merged to defaulted to separate.

bradygaster commented:
I think it'd be ideal if it were all in one place. As in, if you had two top-level folders, one of which contained the API and SignalR Hub and supporting "stuff," and then the second top-level folder contained this repo's contents.

h5aaimtron commented:
I think I can manage that tonight. Do you have a preferred repo name and/or top level folder name preference for the 2 folders?

bradygaster commented:
Nothing fancy - "client" and "server" are fine. :)

h5aaimtron commented:
https://github.com/h5aaimtron/sample_chatWithIdentity

Cleaned it up a little bit. Let me know if there is any additional work that needs to be done.

bradygaster commented:
really good work on the readme and re-organization. cloning and trying it out now - sorry for the delay, was wrapped up in a few other items yesterday.

h5aaimtron commented:
Thank you and no worries. I really like .NET Core and SignalR specifically, so I wanted to give back a little bit. If there are any questions or anything unclear, let me know.

bradygaster commented:
@h5aaimtron - I tried things out. Followed instructions and immediately noticed this error in the client side. Very close! I'd also advise you draw out the logged-in user's name. Would make it more evident that it's multiple users for those who want to pop open multiple browsers to give it a spin. Here's the error i'm seeing when I run both the server and the client. 

![image](https://user-images.githubusercontent.com/41929050/61896373-b3d12b00-aec9-11e9-8dc9-13994b11a6f1.png)

h5aaimtron commented:
@bradygaster I'll look over the code and make the error messages more reasonable. I suspect that the account wasn't created and therefor the token endpoint actually returned a unauthorized error that the client interpreted incorrect. I'll tag you once the changes are committed.

h5aaimtron commented:
@bradygaster I updated the sample to return better error messages when registering or signing in. They aren't perfect, but will tell you if a problem happened. I also added information to the readme that you would need to setup a SQL server or SQL Express connection in the appsettings.json so that the necessary identity tables can be generated. I'm not seeing any errors on my end when running it at this point, but will await your confirmation.

bradygaster commented:
Sounds good @h5aaimtron - sorry i just caught this thread, had been heads-down on a few other tasks but will dive back in to this tonight and get you feedback by end-of-week. thanks for the persistence and passion!

h5aaimtron commented:
@bradygaster wanted to check in on this as its been a few months. If there is anything you need on this sample, let me know, or if you're looking for other complex examples, I may be of help. I'm personally using SignalR Core 2.2 in a personal project similar to Discord with voice chat. I'm hoping to release it in a few months and will likely do a sample app with a simple WebRTC mesh network to put out there.
